### PR TITLE
[Feature] aten::Relabel_() for the GPU

### PR DESCRIPTION
**Pull request by yaox12**
## Description

This implements `aten:Relabel_()` for the GPU, which thus enables `g.edge_subgraph()/in_subgraph()/out_subgraph()` with `relabel_node=True`, as well as `EdgeDataLoader(negative_sampler=None)`, run on the GPU.
The ordered hashtable in CUDA is used for node mapping and relabeling.

Fix this issue https://github.com/dmlc/dgl/issues/3216.

## Checklist

- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
- Modify `src/array/array.cc`: enable switch for the GPU version
- Modify `src/array/cuda/array_op_impl.cu`: CUDA implementation of `Relabel_()`
- Modify `src/graph/heterograph.cc`: enable `g.edge_subgraph()` with `preserve_nodes=False` for graphs on the GPU
- Modify `tests/cpp/test_aten.cc`: enable the unit test for `aten::Relabel_()` on the GPU
- Modify `tests/compute/test_subgraph.py`: enable unit tests for `g.edge_subgraph()/in_subgraph()/out_subgraph()` on the GPU

### CODE REVIEW

File: src/array/cuda/array_op_impl.cu [hunk 1]
```diff
@@ -258,6 +260,84 @@ IdArray Range(IdType low, IdType high, DLContext ctx) {
...
+  device->StreamSync(ctx, thr_entry->stream);
+  device->FreeWorkspace(ctx, num_induced_device);
+
+  // resize the induced nodes
+  induced_nodes->shape[0] = num_induced;
+
+  // relabel
+  for (IdArray arr : arrays) {
+    const int64_t length = arr->shape[0];
+    int nt = cuda::FindNumThreads(length);
```
Comment (nv-dlasalle) [inline src/array/cuda/array_op_impl.cu:328]: I think we should stay away from `cuda::FindNumThreads()`, as it tries to use the maximum number of threads per block, which can lead insufficient use of available SMs, as well as performance anomalies as the number of threads changes with the input size. I'd advice using a fixed number of threads (128 or 256).

**Review by nv-dlasalle — APPROVED** [review]
Overall looks good to me.

File: src/array/cuda/array_op_impl.cu (hunk 1 above)
Comment (yaox12) [reply src/array/cuda/array_op_impl.cu]: Fixed f8d4bb8.

**Comment by nv-dlasalle** [comment]
@BarclayII Are you satisfied with the enabled unit tests, or do you want any other changes before merging?

**Review by BarclayII — APPROVED** [review]
Thanks a lot!  Appreciate it.

**Pull request merged**